src/CityOnPlanet.cpp

```diff
@@ -85,69 +86,179 @@ void CityOnPlanet::RemoveStaticGeomsFromCollisionSpace()
+
+	out.rarityAirless = buildingDef.value("airless-rarity", 1.0);
+	out.rarityAtmo = buildingDef.value("atmo-rarity", 1.0);
+
+	std::string kind = buildingDef.value("kind", "");
+
+	out.buildingKind = SectorKind::None;
+	if (kind == "storage")
+		out.buildingKind = SectorKind::Storage;
+	else if (kind == "industry")
+		out.buildingKind = SectorKind::Industry;
+	else if (kind == "monument")
+		out.buildingKind = SectorKind::Monument;
+	else if (kind == "habitat")
+		out.buildingKind = SectorKind::Habitat;
+	else if (kind == "frontier")
+		out.buildingKind = SectorKind::Frontier;
+
+	std::string idleName = buildingDef.value("idle", "");
+	if (!idleName.empty()) {
+		out.idleAnimation = out.model->FindAnimation(idleName);
+
+		if (out.idleAnimation == nullptr) {
+			Log::Warning("\tIdle animation {} specified for building {} not found!", idleName, key);
 		}
+	} else {
+		out.idleAnimation = out.model->FindAnimation("idle");
 	}
+
+	Log::Verbose("\tLoaded city building {} ({}x{}c, atm: {}, arl: {}, kind: {}, idleAnim: {})",
+		key, out.cellSize[0], out.cellSize[1], out.rarityAtmo, out.rarityAirless, out.buildingKind, out.idleAnimation != nullptr);
 }
 
-//static
-void CityOnPlanet::LookupBuildingListModels(citybuildinglist_t *list)
+void CityOnPlanet::LoadCityFlavour(const FileSystem::FileInfo &file)
 {
-	std::vector<Model *> models;
+	Json fileData = JsonUtils::LoadJson(file.Read());
+	if (!fileData.is_object()) {
+		Log::Info("CityOnPlanet: Could not load city definition file '{}' as a valid JSON file.", file.GetPath());
+		return;
+	}
 
-	//get test newmodels - to be replaced with building set definitions
-	{
-		std::set<std::string> filenames; // set so we get unique names
-		EnumerateNewBuildings(filenames);
-		for (auto it = filenames.begin(), itEnd = filenames.end(); it != itEnd; ++it) {
-			// find/load the model
-			Model *model = Pi::modelCache->FindModel(*it);
-
-			// good to use
-			models.push_back(model);
+	Log::Info("Loading city definition file '{}'", file.GetPath());
+
+	const Json &sizeDef = fileData["size"];
+	if (!sizeDef.is_array()) {
+		Log::Warning("\tMissing city size definition array! Should be \"size\": [ ... ]");
+		return;
+	}
+
+	const Json &buildingDefs = fileData["buildings"];
+	if (!buildingDefs.is_object()) {
+		Log::Warning("\tMissing city buildings object! Should be \"buildings\": { ... }");
+		return;
+	}
+
+	CityFlavourType cityType = {};
+
+	uint32_t numSizeDefs = 0;
+	for (const Json &item : sizeDef) {
+		CityRadiusDef radius = {};
+		numSizeDefs++;
+
+		radius.population = item.value("population", 0.0);
+		radius.baseSize = item.value("baseSize", 0.0);
+		radius.atmoSize = item.value("atmoSize", 0.0);
+		radius.randomSize = item.value("randomSize", 0.0);
+		radius.density = item.value("density", 1.0);
+
+		if (!cityType.sizeDefs.empty() && radius.population > cityType.sizeDefs.back().population) {
+			Log::Warning("\tSize definition {} has higher population threshold than previous ({}b vs. {}b); ignoring",
```

Comment:
Why not just sort it later?
Оr at least write a warning in `default.json` I think.

Reply:
The console and log file are a good enough warning as far as I'm considered, though I do eventually want to capture the warning messages and display them in the menu at startup... I've added this constraint here somewhat arbitrarily during development to avoid complicating things too much, but it's much better to keep the out-of-order declaration and sort. Will address that.